File: backend/app/schemas/ml.py
"""
Pydantic schemas for ML endpoints.
"""
from pydantic import BaseModel, Field
from datetime import datetime


class DemandForecastResponse(BaseModel):
    """Response schema for demand forecast."""
    product_id: str
    product_name: str
    horizon: int = Field(..., description="Forecast horizon in days")
    daily_predictions: list[float] = Field(..., description="Daily demand predictions")
    predicted_total_demand: float
    model_type: str = Field(..., description="'ml' or 'baseline'")
    current_stock: int
    reorder_point: int
    lead_time_days: int
    recommended_reorder_quantity: float
    safety_stock: float
    training_timestamp: str | None = None
    metrics: dict = Field(default_factory=dict)


class ModelMetricsResponse(BaseModel):
    """Response schema for model metrics."""
    model_trained: bool
    model_type: str | None = None
    trained_at: str | None = None
    feature_names: list[str] = Field(default_factory=list)
    metrics: dict = Field(default_factory=dict)
    message: str | None = None


class TrainModelRequest(BaseModel):
    """Request schema for model training."""
    force_retrain: bool = Field(False, description="Force retraining even if model exists")


class TrainModelResponse(BaseModel):
    """Response schema for model training."""
    status: str
    model_type: str | None = None
    trained_at: str | None = None
    metrics: dict = Field(default_factory=dict)

File: frontend/src/api/ml.ts
/**
 * ML API — Demand Forecasting Endpoints
 */
import { apiClient } from './client'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface DemandForecast {
  product_id: string
  product_name: string
  horizon: number
  daily_predictions: number[]
  predicted_total_demand: number
  model_type: 'ml' | 'baseline'
  current_stock: number
  reorder_point: number
  lead_time_days: number
  recommended_reorder_quantity: number
  safety_stock: number
  training_timestamp: string | null
  metrics: {
    val_mae?: number
    val_rmse?: number
    baseline_mae?: number
    baseline_rmse?: number
    n_train?: number
    n_val?: number
  }
}

export interface ModelMetrics {
  model_trained: boolean
  model_type?: string
  trained_at?: string
  feature_names: string[]
  metrics: {
    val_mae?: number
    val_rmse?: number
    baseline_mae?: number
    baseline_rmse?: number
    n_train?: number
    n_val?: number
  }
  message?: string
}

// ---------------------------------------------------------------------------
// API Functions
// ---------------------------------------------------------------------------

/**
 * Get demand forecast for a specific product.
 */
export async function getDemandForecast(
  productId: string,
  horizon: number = 7
): Promise<DemandForecast> {
  const { data } = await apiClient.get<DemandForecast>(
    `/ml/demand-forecast/${productId}`,
    { params: { horizon } }
  )
  return data
}

/**
 * Get current model metrics and metadata.
 */
export async function getModelMetrics(): Promise<ModelMetrics> {
  const { data } = await apiClient.get<ModelMetrics>('/ml/metrics')
  return data
}

/**
 * Train or retrain the global forecasting model (MANAGER/ADMIN only).
 */
export async function trainModel(forceRetrain: boolean = false): Promise<{
  status: string
  model_type?: string
  trained_at?: string
  metrics: Record<string, unknown>
}> {
  const { data } = await apiClient.post('/ml/train', {
    force_retrain: forceRetrain,
  })
  return data
}
